pkg/certificate/mrc_reconciler.go

```diff
@@ -0,0 +1,164 @@
+package certificate
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	"github.com/openservicemesh/osm/pkg/constants"
+)
+
+// UseCase is how the certificate is used. Each use case will have its own rotation mechanism, that is necessary to
+// determine if we can propagate the rotation forward.
+type UseCase string
+
+const (
+	// UseCaseValidatingWebhook maps to the certificate use case for the validating webhook.
+	UseCaseValidatingWebhook UseCase = "validatingWebhook"
+
+	// UseCaseMutatingWebhook maps to the certificate use case for the Mutating Webhook
+	UseCaseMutatingWebhook UseCase = "mutatingWebhook"
+
+	// UseCaseXDSControlPlane maps to the certificate use case for the XDS Control Plane
+	UseCaseXDSControlPlane UseCase = "xdsControlPlane"
+
+	// UseCaseSidecar maps to the certificate use case for the  sidecar service certs
+	UseCaseSidecar UseCase = "sidecar"
+
+	// UseCaseBootstrap maps to the certificate use case for the  Bootstrap secrets
+	UseCaseBootstrap UseCase = "bootstrap"
+
+	// UseCaseGateway maps to the certificate use case for the  Gateway ingress cert.
+	UseCaseGateway UseCase = "gateway"
+)
+
+func (uc UseCase) String() string {
+	return string(uc)
+}
+func (m *Manager) handleMRCEvent(event MRCEvent) error {
+	mrc := event.MRC
+	// TODO(5046): improve logging in this method.
+	if shouldUpdateMRCComponentStatus(mrc) {
+		if err := m.updateMRCComponentStatus(mrc); err != nil {
+			return err
+		}
+	} else if !isTerminal(mrc) {
+		if err := m.updateMRCState(mrc); err != nil {
+			return err
+		}
+	} else if m.shouldSetIssuers(mrc) {
+		log.Info().Msgf("setting new certificate issuers on the Certificate Manager for MRC %s in stage %s", mrc.GetName(), mrc.Status.State)
+		return m.setIssuers(mrc)
+	}
+	return nil
+}
+
+// Component statuses determine if we're ready to transition the mrc to the next state.
+// TODO(5046): add unit tests for this, and other methods in here.
+func shouldUpdateMRCComponentStatus(mrc *v1alpha2.MeshRootCertificate) bool {
+	if isTerminal(mrc) {
+		return false
+	}
+
+	if len(mrc.Status.Conditions) == 0 {
+		return false
+	}
+
+	transitionAfter := mrc.Status.TransitionAfter
+	if transitionAfter == nil {
+		return false
+	}
+
+	//The truncate is needed so we truncate the monotonic clock.
+	return transitionAfter.Truncate(0).After(time.Now())
+}
+
+// update the MRCComponentStatus
+func (m *Manager) updateMRCComponentStatus(mrc *v1alpha2.MeshRootCertificate) error {
+	var state v1alpha2.MeshRootCertificateComponentStatus
+	switch mrc.Status.State {
```

Comment:
You can't use state like this; we talked about how both `state` and conditions shouldn't be set until all component statuses agree.

Reply:
that was just a variable,  i probably shouldn't have named it state. renamed to status